src/utils.ts

```diff
@@ -0,0 +1,456 @@
+      const typedDecodedScript: DecodedScript = {
+        type: input.decoded.type as string,
+        address: input.decoded.address as string,
+        timelock: input.decoded.timelock ? input.decoded.timelock as number : null,
+        value: input.decoded.value ? input.decoded.value as number : null,
+        tokenData: input.decoded.token_data ? input.decoded.token_data as number : null,
+      };
+      const typedInput: Input = {
+        txId: input.tx_id as string,
+        index: input.index as number,
+        value: input.value as number,
+        tokenData: input.token_data as number,
+        script: input.script as string,
+        decoded: typedDecodedScript,
+        token: input.token as string,
+      };
+
+      return typedInput;
+    }),
+    outputs: tx.outputs.map((output): Output => {
+      const typedDecodedScript: DecodedScript = {
+        type: output.decoded.type as string,
+        address: output.decoded.address as string,
+        timelock: output.decoded.timelock ? output.decoded.timelock as number : null,
+        value: output.decoded.value ? output.decoded.value as number : null,
+        tokenData: output.decoded.token_data ? output.decoded.token_data as number : null,
+      };
+
+      const typedOutput: Output = {
+        value: output.value as number,
+        tokenData: output.token_data as number,
+        script: output.script as string,
+        decoded: typedDecodedScript,
+        token: output.token as string,
+        spentBy: output.spent_by ? output.spent_by as string : null,
+      };
+
+      return typedOutput;
+    }),
+    parents: tx.parents,
+    tokens: tx.tokens.map((token: any): Token => {
+      const parsedToken: Token = {
+        uid: token.uid as string,
+        name: token.name as string,
+        symbol: token.symbol as string,
+      };
+
+      return parsedToken;
+    }),
+    raw: tx.raw as string,
+  };
+
+  return parsedTx;
+};
+
+export const getBlockByTxId = async (txId: string, noCache: boolean = false) => {
+  return downloadTx(txId, noCache);
+};
+
+export async function* syncToLatestBlock(): AsyncGenerator<StatusEvent> {
+  const ourBestBlock: Block = await getWalletServiceBestBlock();
+  const fullNodeBestBlock: Block = await getFullNodeBestBlock();
+
+  // Check if our best block is still in the fullnode's chain
+  const ourBestBlockInFullNode = await getBlockByTxId(ourBestBlock.txId, true);
+
+  if (!ourBestBlockInFullNode.success) {
+    yield {
+      type: 'error',
+      success: false,
+      message: 'Could not validate our best block',
+      error: ourBestBlockInFullNode.message,
+    };
+
+    return;
+  }
+
+  const { meta } = ourBestBlockInFullNode;
+
+  if ((meta.voided_by &&
+       meta.voided_by.length &&
+       meta.voided_by.length > 0)) {
+    yield {
+      type: 'reorg',
+      success: false,
+      message: 'Our best block was voided, we should reorg.',
+    };
+
+    return;
+  }
+
+  if (ourBestBlock.height > meta.height) {
```

Comment:
I didn't understand when this situation may happen. The txId in the full node has a height that is bigger than the same txId in the wallet service?

Reply:
We are downloading the best block from the full node and comparing it to our own. If a reorg happened, I believe there is a chance that the full node's best block height is lower than ours

Not sure if this is a real situation tho, this is one of the things I want to test using the full node simulator when it is available @msbrogli

Reply:
It should never happen. Anyway, we should assert `outBestBlock.height <= meta.height` (or just verify and raise an exception as we are doing here).